### ISSUE TITLE
Jobs keep re-processing without ever starting

### ISSUE THREAD

**Issue by adsamcik** [sergeyzwezdin/Hangfire.Mongo#50]
I don't know if I've done something wrong but it worked before I changed from sql to mongo. 

What exactly happens is that job gets queued and starts processing and starts processing and starts processing and so on. So in the history there can be tens of processing states but it seems to never finish or fail. I put a bunch of WriteLines to find how are the processes executed and where they end and it never showed a single WriteLine. So I guess the jobs never even start. I am running on ASP.NET Core 1.1. 

Also when deleting that job, it stays in the queue.

edit:// Did a quick switch to sql and everything works as expected. Jobs didn't stay in queue and got processed properly.

**Comment by maouida** [comment]
I'm experiencing a similar issue. The jobs stays in `Enqueued` state and never get executed. When I switch to SqlServer, it works correctly.

**Comment by gottscj** [comment]
Hi @adsamcik and @maouida,

Thank you for your feedback, I have created a small .netcore app and cannot reproduce what your are seeing.
Could I get you to try it out?
(Even though Github says it allows .zip, I was unsuccessful in uploading this, so I put it on my onedrive)

https://1drv.ms/u/s!AqZWS5rQI9zmsGe6A5S9fmNrjKeB

Also could you add some more debug info?

Thanks
Jonas

**Comment by adsamcik** [comment]
@gottscj I don't have mongodb locally so I couldn't test yours. but I have created new asp.net core 1.1 project from scratch and included at the end of configure 
`JobStorage.Current = new MongoStorage("mongodb://localhost", "mymongotest");
            using (var server = new BackgroundJobServer()) {
                for (var i = 0; i < 10; i++) {
                    BackgroundJob.Enqueue(() => Console.WriteLine($"Fire-and-forget ({i})"));
                }
                Console.ReadKey(true);
            }`
It waited for a minute or two but nothing was happening. Update: I tested yours code too in the end and it didn't work either. Both projects were tested on Ubuntu 16.04 so that might be the problem.

Collection: hangfire.jobQueue
> { "_id" : 1, "JobId" : 1, "Queue" : "default", "FetchedAt" : null }

Collection: hangfire.set


> {
        "_id" : ObjectId("58457110ae20fd66382cdc09"),
        "JobId" : 1,
        "Name" : "Processing",
        "Reason" : null,
        "CreatedAt" : ISODate("2016-12-05T13:52:16.233Z"),
        "Data" : "{\"StartedAt\":\"2016-12-05T13:52:16.2262210Z\",\"ServerId\":\"65c6187b-4da8-4699-a668-53f35443f4ee\",\"WorkerId\":\"0696750e-5253-4e46-a7d0-10a9f1711e69\"}"
}
{
        "_id" : ObjectId("58457111ae20fd66382cdc0a"),
        "JobId" : 1,
        "Name" : "Processing",
        "Reason" : null,
        "CreatedAt" : ISODate("2016-12-05T13:52:17.234Z"),
        "Data" : "{\"StartedAt\":\"2016-12-05T13:52:17.2284720Z\",\"ServerId\":\"65c6187b-4da8-4699-a668-53f35443f4ee\",\"WorkerId\":\"fc4c2850-9518-444b-aa96-2a01aa8ae289\"}"
}
{
        "_id" : ObjectId("58457114ae20fd66382cdc0b"),
        "JobId" : 1,
        "Name" : "Processing",
        "Reason" : null,
        "CreatedAt" : ISODate("2016-12-05T13:52:20.200Z"),
        "Data" : "{\"StartedAt\":\"2016-12-05T13:52:20.1949060Z\",\"ServerId\":\"65c6187b-4da8-4699-a668-53f35443f4ee\",\"WorkerId\":\"007bec67-1b6f-4869-be17-4092a4b33584\"}"
}



Collection: hangfire.job
> {
        "_id" : 1,
        "StateId" : ObjectId("58457114ae20fd66382cdc0b"),
        "StateName" : "Processing",
        "InvocationData" : "{\"Type\":\"System.Console, System.Console, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a\",\"Method\":\"WriteLine\",\"ParameterTypes\":\"[\\\"System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e\\\"]\",\"Arguments\":\"[\\\"\\\\\\\"Fire-and-forget (0)\\\\\\\"\\\"]\"}",
        "Arguments" : "[\"\\\"Fire-and-forget (0)\\\"\"]",
        "CreatedAt" : ISODate("2016-12-05T13:50:50.087Z"),
        "ExpireAt" : null
}

**Comment by MarLoe** [comment]
@adsamcik, runnig on Ubuntu might explain the problem :)

I do not have much Ubuntu experience, but I am in the process of spinning a virtual Ubuntu up and try it out - to see if I can reproduce what you are seeing.

In the meantime, it would be great with any info you can provide us. Stack traces, debug info etc.

I will get back to you, once I have my Ubuntu running successfully.

**Comment by adsamcik** [comment]
@MarLoe the program you posted doesn't generate any console output and I don't really have time at this time of semester to dig deeper :) Running it on ubuntu is simple, just follow the dotnet tutorial, paste the source file in there and run it like on windows if you use console. You should be ready in couple of minutes.

**Comment by MarLoe** [comment]
@adsamcik, I finally have an Ubuntu up and running with .NET Core. And I also see the problem you are experiencing. I have also identified the problem. But it a matter of platform unsupported APIs :(

I am looking into alternatives, but there is no quick fix to the problem. I will get back once I have some news.

**Comment by adsamcik** [comment]
@MarLoe At least the issue was identified :)

**Comment by MarLoe** [comment]
@adsamcik, I have made a branch "netcore_unix_support". If you have the time, please download this and try it out. I have included a .NETCore sample project (the one previously attached to this thread), that would allow you to test it easily.

I have only tested this branch on Ubuntu, so I will need to test it on Windows before making a pull request to master.

Enjoy,
Martin

**Comment by MarLoe** [comment]
@adsamcik & @maouida, made a pull request #53 that should fix your problem. A new NuGet release of hangfire.Mongo will be released once #53 is merged into master.

**Comment by adsamcik** [comment]
Had a busy week. Tested it right now and it seems to work fine. Found few bugs (or differences from how sql works) but nothing too function breaking. Thanks for the quick patch :)

**Comment by MarLoe** [comment]
@adsamcik, you're welcome! 
Please let us know of your findings between this and the sql implementation. You could create a new issue named e.g. Hangfire.Mongo vs Hangfire.Sql and start listing your findings there.
